src/partition/repart.c

```diff
@@ -7392,6 +7435,60 @@ static int context_crypttab(Context *context) {
         return 0;
 }
 
+/* update block sizes for verity siblings, calculate hash partition size if requested */
+static int context_update_verity_size(Context *context) {
+        int r;
+
+        assert(context);
+
+        LIST_FOREACH(partitions, p, context->partitions) {
+                Partition *dp;
+
+                if (p->verity != VERITY_HASH)
+                        continue;
+
+                if (p->dropped)
+                        continue;
+
+                if (PARTITION_EXISTS(p)) /* Never format existing partitions */
+                        continue;
+
+                assert_se(dp = p->siblings[VERITY_DATA]);
+
+                if (p->verity_data_block_size == UINT64_MAX)
+                        p->verity_data_block_size = context->fs_sector_size;
+
+                if (p->verity_hash_block_size == UINT64_MAX)
+                        p->verity_hash_block_size = context->fs_sector_size;
+
+                uint64_t sz;
+                if (dp->size_max != UINT64_MAX) {
+                        r = calculate_verity_hash_size(
+                                        dp->size_max,
+                                        p->verity_hash_block_size,
+                                        p->verity_data_block_size,
+                                        &sz);
+                        if (r < 0)
+                                return log_error_errno(r, "Failed to caculate size of dm-verity hash partition: %m");
+
+                        if (sz > p->size_max)
+                                log_warning("SizeMaxBytes=%s of dm-verity hash partition too small for data partition "
+                                            "with SizeMaxBytes=%s. Hash partition would require %s bytes for a data "
+                                            "partition of specified max size. Consider increasing SizeMaxBytes= of "
+                                            "hash partition, or decreasing SizeMaxBytes= of data partition.",
+                                            FORMAT_BYTES(p->size_max), FORMAT_BYTES(dp->size_max), FORMAT_BYTES(sz));
+                        else if (p->size_min == UINT64_MAX) {
```

Comment:
Maybe, also update when the specified minimum size is smaller than the derived one?

Reply:
Hmm, wouldn't this be another case like above, were only a warning should be emitted?
The _actual_ hash data may potentially fit into the user-defined minimum size

Reply:
OK. Sorry for the noise. Thanks.

Reply:
What do you think of including it in the warning like this?

```diff
--- a/src/partition/repart.c
+++ b/src/partition/repart.c
@@ -7471,12 +7471,12 @@ static int context_update_verity_size(Context *context) {
                         if (r < 0)
                                 return log_error_errno(r, "Failed to caculate size of dm-verity hash partition: %m");

-                        if (sz > p->size_max)
-                                log_warning("SizeMaxBytes=%s of dm-verity hash partition too small for data partition "
-                                            "with SizeMaxBytes=%s. Hash partition would require %s bytes for a data "
-                                            "partition of specified max size. Consider increasing SizeMaxBytes= of "
-                                            "hash partition, or decreasing SizeMaxBytes= of data partition.",
-                                            FORMAT_BYTES(p->size_max), FORMAT_BYTES(dp->size_max), FORMAT_BYTES(sz));
+                        if (sz > p->size_min || sz > p->size_max)
+                                log_warning("The dm-verity hash partition %s may be too small for a data partition "
+                                            "with SizeMaxBytes=%s. The hash partition would require %s for a data "
+                                            "partition of specified max size. Consider increasing the size of the "
+                                            "hash partition, or decreasing SizeMaxBytes= of the data partition.",
+                                            dp->definition_path, FORMAT_BYTES(dp->size_max), FORMAT_BYTES(sz)));
                         else if (p->size_min == UINT64_MAX) {
                                 log_debug("Derived size %s of verity hash partition %s from verity data partition %s.",
                                                 FORMAT_BYTES(sz), p->definition_path, dp->definition_path);
```

Reply:
Looks better for me.